pkg/apimachinery/registered/registered.go

```diff
@@ -114,22 +114,24 @@ func IsEnabledVersion(v unversioned.GroupVersion) bool {
 	return found
 }
 
-// EnabledVersions returns all enabled versions.
-func EnabledVersions() (ret []unversioned.GroupVersion) {
-	for v := range enabledVersions {
-		ret = append(ret, v)
+// EnabledVersions returns all enabled versions.  Groups are randomly ordered, but versions within groups
+// are priority order from best to worst
+func EnabledVersions() []unversioned.GroupVersion {
+	ret := []unversioned.GroupVersion{}
+	for _, groupMeta := range groupMetaMap {
+		ret = append(ret, groupMeta.GroupVersions...)
```

Comment:
I agree on the version ordering.  I'm just trying to understand why enabled
versions is random.  It's fine for this issue, but I don't know that I
consider randomness correct in this context.

On Wed, Feb 10, 2016 at 8:31 AM, David Eads <notifications@github.com>
wrote:

> In pkg/apimachinery/registered/registered.go
> <https://github.com/kubernetes/kubernetes/pull/20846#discussion_r52456327>
> :
>
> > @@ -114,22 +114,24 @@ func IsEnabledVersion(v unversioned.GroupVersion) bool {
> >  	return found
> >  }
> >
> > -// EnabledVersions returns all enabled versions.
> > -func EnabledVersions() (ret []unversioned.GroupVersion) {
> > -	for v := range enabledVersions {
> > -		ret = append(ret, v)
> > +// EnabledVersions returns all enabled versions.  Groups are randomly ordered, but versions within groups
> > +// are priority order from best to worst
> > +func EnabledVersions() []unversioned.GroupVersion {
> > +	ret := []unversioned.GroupVersion{}
> > +	for _, groupMeta := range groupMetaMap {
> > +		ret = append(ret, groupMeta.GroupVersions...)
>
> Why are groups randomly ordered? What value does it provide?
>
> When choosing how to encode with a legacy encoder, the first matching
> group is used. This is a natural method for getting all the GroupVersions
> and since the group will be traversed eventually, the group ordering
> doesn't matter. The version ordering within that group is critical.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20846/files#r52456327>.
>